### ISSUE TITLE
[Bug]: flatpak-builder --extra-sources not working ?

### ISSUE THREAD

**Issue by TTimo** [flatpak/flatpak-builder#499]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a bug that matches the one I want to file, without success.
- [X] If this is an issue with a particular app, I have tried filing it in the appropriate issue tracker for the app (e.g. under https://github.com/flathub/) and determined that it is an issue with Flatpak itself.
- [X] This issue is not a report of a security vulnerability (see [here](https://github.com/flatpak/flatpak/blob/main/SECURITY.md) if you need to report a security issue).

### Flatpak version

1.15.0

### What Linux distribution are you using?

Manjaro Linux

### Linux distribution version

Linux phoenix 5.19.17-2-MANJARO flatpak/flatpak#1 SMP PREEMPT_DYNAMIC Sun Nov 6 00:08:27 UTC 2022 x86_64 GNU/Linux

### What architecture are you using?

x86_64

### How to reproduce

I have the following module:

```
            {
                "type": "archive",
                "url": "https://netix.dl.sourceforge.net/project/scons/scons/3.0.1/scons-3.0.1.tar.gz",
                "sha256": "24475e38d39c19683bc88054524df018fe6949d70fbd4c69e298d39a0269f173"
            },
```

I run this command:

`flatpak-builder --extra-sources=/home/timo/GtkRadiant/extra-sources --force-clean --disable-updates --repo=gtkradiant-repo gtkradiant io.github.TTimo.GtkRadiant.json`

A directory with a local copy of the tarball is setup, the checksum matches:

```
timo@phoenix ~/G/extra-sources> pwd
/home/timo/GtkRadiant/extra-sources
timo@phoenix ~/G/extra-sources> ls -1al
total 632
drwxr-xr-x  3 timo timo   4096 Nov 24 20:01 ./
drwxr-xr-x  4 timo timo   4096 Nov 24 20:01 ../
drwxr-xr-x 15 timo timo   4096 Nov 24 19:51 gtkradiant/
-rw-r--r--  1 timo timo 634815 Nov 14  2017 scons-3.0.1.tar.gz
timo@phoenix ~/G/extra-sources> sha256sum scons-3.0.1.tar.gz
24475e38d39c19683bc88054524df018fe6949d70fbd4c69e298d39a0269f173  scons-3.0.1.tar.gz
```

### Expected Behavior

flatpak-builder should be using my local tarball

### Actual Behavior

flatpak-builder always downloads from the URL instead

### Additional Information

It's possible I misunderstand the documentation, I haven't found any working examples online of using --extra-sources

**Comment by smcv** [comment]
Flatpak isn't responsible for reading manifests, issue moved to flatpak-builder.

**Comment by retiredfeline** [comment]
Doesn't work for me too:

openSUSE 15.4
Flatpak 1.12.8

Command: flatpak-builder --extra-sources=/home/data/software/uC/8088 build-dir io.github.gcc_ia16.Gcc_ia16_2204.yaml

Relevant manifest lines:

```
    sources:
      - type: archive
        url: https://launchpad.net/~tkchia/+archive/ubuntu/build-ia16/+sourcefiles/gcc-ia16-elf/6.3.0-20230219.07-ppa230219074~jammy/gcc-ia16-elf_6.3.0-20230219.07.orig.tar.xz
        sha256: "a8db1ea5603ed7d1c832ee452f3d7fa1093e00aa194146b6b15db55b830d3d15"
```

Doesn't use the local tarball, starts fetching from the remote site with curl.

**Comment by TiZ-HugLife** [comment]
Yeah, this is broken for me, too. I'm trying to do a local build for an application that wants Lua, and Lua's site is currently down for maintenance. I already downloaded the tarball, but it won't use it.

```
❯ org.flatpak.Builder --extra-sources=$(pwd) --force-clean --ccache --user --install build org.geany.Geany.yml
Downloading sources
Fetching git repo https://gitlab.gnome.org/GNOME/vte.git, ref refs/tags/0.72.4
remote: Enumerating objects: 1, done.
remote: Counting objects: 100% (1/1), done.
remote: Total 1 (delta 0), reused 1 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (1/1), 269 bytes | 269.00 KiB/s, done.
Fetching LFS assets
fetch: Fetching all references...
Downloading https://www.lua.org/ftp/lua-5.1.5.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:02 --:--:--     0^C
```

Fortunately, Lua is a shared module, and you can work around the site outage by switching `https://www.lua.org` with `https://www.tecgraf.puc-rio.br/lua/mirror` in your local copy of the shared module. Mirror sourced [from this comment on /r/lua.](https://www.reddit.com/r/lua/comments/1di6evb/luaorg_site_is_down/l9201of/)

**Comment by MoralCode** [comment]
Workaround: if you have a file downloaded from `https://example.com/my-zip.zip` with `dest-filename` set to `zip.zip` and a hash of `abcd`, copy (or symlink) your pre-downloaded file into `.flatpak-builder/downloads/abcd/zip.zip`

**Comment by bbhtt** [comment]
It works but you have to recreate the exact directory setup that flatpak-builder uses:

```
pwd                      
/home/bbhtt/Downloads
~/Downloads 
→ tree -L 3 
.
└── downloads
    └── 6940718dfc3eff4258203ad5021090933e5c04707d5ca8cc9e73c94a7894ea9f
        └── pathlib-1.0.1.tar.gz

3 directories, 1 file
```

Which I guess doesn't match the common user expectation that they can just dump files/archives in any directory, pass that as `--extra-sources` and it works.

Which also makes the option not very useful.

But looking at the code 8 years ago, it seems like this precondition of recreating the directory structure was always there.

https://github.com/flatpak/flatpak-builder/commit/894311e24353f406771d1cc35c108aad9a33cd31#diff-2dd9f34adefc6092a37f176f14dea5bc8c7885a3b23d06ad1d9a0a8b1dda4aafR199-R202

**Comment by bbhtt** [comment]
ah the commit message says that already:

> Those source directories do need the same
structure as the ones that flatpak-builder
creates during processing in .flatpak-builder
and which is also used in the exported sources
runtime.

**Comment by bbhtt** [comment]
So this particular issue is "works as expected", I will add it to the documentation about the directory setup and close it.

Someone can provide a PR to improve the behaviour of this argument, but if nothing comes out we should probably just remove it.
